### PR TITLE
add integration test for copy from with low bulk size

### PR DESCRIPTION
**Pull request by m-kharbat**

### CODE REVIEW

File: sql/src/main/java/io/crate/operation/projectors/BatchIteratorBackpressureExecutor.java [hunk 1]
```diff
@@ -154,17 +154,7 @@ private void unsafeConsumeIterator(BatchIterator batchIterator) {
                     if (tryExecuteBulk(false) == false) {
                         collectingEnabled.set(false);
                         if (isScheduledCollectionRunning.compareAndSet(false, true)) {
-                            try {
```
Comment (andreidan) [inline sql/src/main/java/io/crate/operation/projectors/BatchIteratorBackpressureExecutor.java]: please drop this change as it's done in another PR

File: sql/src/main/java/io/crate/operation/projectors/ShardingUpsertExecutor.java [hunk 2]
```diff
@@ -159,6 +159,7 @@ private BooleanSupplier createBackpressureTrigger(NodeJobsCounter nodeJobsCounte
             for (ShardLocation shardLocation : requestsByShard.keySet()) {
                 String requestNodeId = shardLocation.nodeId;
                 if (nodeJobsCounter.getInProgressJobsForNode(requestNodeId) >= MAX_NODE_CONCURRENT_OPERATIONS) {
```
Comment (andreidan) [inline sql/src/main/java/io/crate/operation/projectors/ShardingUpsertExecutor.java]: shouldn't we try to lower the MAX_NODE_CONCURRENT_OPERATIONS to 2?

File: sql/src/main/java/io/crate/operation/projectors/ShardingUpsertExecutor.java [hunk 3]
```diff
@@ -302,6 +303,7 @@ private void countdown() {
                 listener,
                 BACKOFF_POLICY
             );
+            LOGGER.debug("Executing retry Listener for request: {}", request);
```
Comment (andreidan) [inline sql/src/main/java/io/crate/operation/projectors/ShardingUpsertExecutor.java]: This log statement is not correct. This should be in scheduleConsumeIterator. (or if we really want to log it in the retry loop we should log it in the command we are passing to the retryListener ie:  l -> { 
LOGGER.debug("Executing retry Listener for request: {}", request);
requestExecutor.execute(request, l);
}..., )

File: sql/src/main/java/io/crate/operation/projectors/ShardingUpsertExecutor.java [hunk 4]
```diff
@@ -159,6 +159,7 @@ private BooleanSupplier createBackpressureTrigger(NodeJobsCounter nodeJobsCounte
             for (ShardLocation shardLocation : requestsByShard.keySet()) {
                 String requestNodeId = shardLocation.nodeId;
                 if (nodeJobsCounter.getInProgressJobsForNode(requestNodeId) >= MAX_NODE_CONCURRENT_OPERATIONS) {
+                    LOGGER.debug("reached maximum concurrent operations for node");
```
Comment (andreidan) [inline sql/src/main/java/io/crate/operation/projectors/ShardingUpsertExecutor.java]: Please also log the node

File: sql/src/test/java/io/crate/integrationtests/ShardingUpsertIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+@ESIntegTestCase.ClusterScope(numDataNodes = 1)
+public class ShardingUpsertIntegrationTest extends SQLTransportIntegrationTest {
+
+    private String copyFilePath = Paths.get(getClass().getResource("/essetup/data/best_practice").toURI()).toUri().toString();
+
+    @Rule
+    public TemporaryFolder folder = new TemporaryFolder();
+
+    public ShardingUpsertIntegrationTest() throws URISyntaxException {
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/ShardingUpsertIntegrationTest.java]: unused, pls remove

File: sql/src/test/java/io/crate/integrationtests/ShardingUpsertIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+import java.net.URISyntaxException;
+import java.nio.file.Paths;
+
+@ESIntegTestCase.ClusterScope(numDataNodes = 1)
+public class ShardingUpsertIntegrationTest extends SQLTransportIntegrationTest {
+
+    private String copyFilePath = Paths.get(getClass().getResource("/essetup/data/best_practice").toURI()).toUri().toString();
+
+    @Rule
+    public TemporaryFolder folder = new TemporaryFolder();
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/ShardingUpsertIntegrationTest.java]: unused, pls remove

File: sql/src/test/java/io/crate/integrationtests/ShardingUpsertIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+    public ShardingUpsertIntegrationTest() throws URISyntaxException {
+    }
+
+    @Test
+    @TestLogging(value = "_root:DEBUG")
+    public void testCopyFromWithLimitedBulkSize() throws Exception {
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/ShardingUpsertIntegrationTest.java]: how can this test result in a retry, if the thread pool size is not adjusted? afaik no retry will ever happen currently

File: sql/src/test/java/io/crate/integrationtests/ShardingUpsertIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    private String copyFilePath = Paths.get(getClass().getResource("/essetup/data/best_practice").toURI()).toUri().toString();
+
+    @Rule
+    public TemporaryFolder folder = new TemporaryFolder();
+
+    public ShardingUpsertIntegrationTest() throws URISyntaxException {
+    }
+
+    @Test
+    @TestLogging(value = "_root:DEBUG")
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/ShardingUpsertIntegrationTest.java]: pls remove, should be only for local debugging

File: sql/src/test/java/io/crate/integrationtests/ShardingUpsertIntegrationTest.java (hunk 5 above)
Comment (m-kharbat) [inline sql/src/test/java/io/crate/integrationtests/ShardingUpsertIntegrationTest.java]: this is needed since toURI() throws URISyntaxException

File: sql/src/test/java/io/crate/integrationtests/ShardingUpsertIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+import org.elasticsearch.test.junit.annotations.TestLogging;
+import org.junit.Test;
+
+import java.net.URISyntaxException;
+import java.nio.file.Paths;
+
+@ESIntegTestCase.ClusterScope(numDataNodes = 1)
+public class ShardingUpsertIntegrationTest extends SQLTransportIntegrationTest {
+
+    private String copyFilePath = Paths.get(getClass().getResource("/essetup/data/best_practice").toURI()).toUri().toString();
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/ShardingUpsertIntegrationTest.java]: as this is used only in one test, pls move it to the test method. this allows you to remove the constructor

File: sql/src/test/java/io/crate/integrationtests/ShardingUpsertIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    @Override
+    protected Settings nodeSettings(int nodeOrdinal) {
+        return Settings.builder()
+            .put(super.nodeSettings(nodeOrdinal))
+            .put("thread_pool.bulk.queue_size", 1)
+            .build();
+    }
+
+    @Test
+    @TestLogging(value = "io.crate.operation.projectors:DEBUG")
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/ShardingUpsertIntegrationTest.java]: pls remove

File: sql/src/main/java/io/crate/operation/projectors/ShardingUpsertExecutor.java [hunk 11]
```diff
@@ -298,7 +288,10 @@ private void countdown() {
 
             listener = new RetryListener<>(
                 scheduler,
-                l -> requestExecutor.execute(request, l),
+                l -> {
+                    LOGGER.debug("Executing retry Listener for request: {}", request);
```
Comment (seut) [inline sql/src/main/java/io/crate/operation/projectors/ShardingUpsertExecutor.java]: would be nice to add the requestNodeId here as well

File: sql/src/main/java/io/crate/operation/projectors/ShardingUpsertExecutor.java [hunk 12]
```diff
@@ -298,7 +288,12 @@ private void countdown() {
 
             listener = new RetryListener<>(
                 scheduler,
-                l -> requestExecutor.execute(request, l),
+                l -> {
+                    LOGGER.debug("Executing retry Listener for request: {} nodeId: {}",
```
Comment (seut) [inline sql/src/main/java/io/crate/operation/projectors/ShardingUpsertExecutor.java]: minor: I'd write nodeId first as this is short and request is probably long
